File: 2020/day1/day1.hs
module Main where

import qualified Data.Set as S
import qualified Data.List as L
import System.IO

main :: IO()
main = do
    contents <- readFile "in.txt"
    let numbers = map (\line -> read line :: Int) $ lines contents
    print $ "Part 1: " ++ show (day1 numbers 2)
    print $ "Part 2: " ++ show (day1 numbers 3)

day1 :: [Int] -> Int -> Int
day1 xs n = product numbers
    where sums = map (\nr -> (nr, sum nr)) $ sequence $ take n $ repeat xs
          numbers = fst $ maybe (take n $ repeat 0, 0) id $ L.find (\x -> (snd x) == 2020) $ sums

File: 2020/day1/day1.exs
defmodule Day1 do
  def day1(n) do
    numbers = File.read!("in.txt")
      |> String.split("\r\n")
      |> Enum.map(&String.to_integer/1)

    if n == 3 do
      combinations = cartesian(numbers, numbers, numbers)
      IO.puts "Part 2:"
      get_product(combinations) |> IO.inspect
    else
      combinations = cartesian(numbers, numbers)
      IO.puts "Part 1:"
      get_product(combinations) |> IO.inspect
    end
  end

  def get_product(combinations) do
    combinations
    |> Enum.map(fn nr -> [nr, Enum.sum(nr)] end)
    |> Enum.find(fn nr -> Enum.at(nr, 1) == 2020 end)
    |> Enum.at(0)
    |> Enum.reduce(fn x, acc -> x * acc end)
  end

  def cartesian(a, b) do
    for x <- a, y <- b, do: [x, y]
  end

  def cartesian(a, b, c) do
    for x <- a, y <- b, z <- c, do: [x, y, z]
  end
end

Day1.day1(2)
Day1.day1(3)
